rfcs/2023-09-19-18604-metrics-data-model-refactor.md

```diff
@@ -0,0 +1,455 @@
+# RFC 18604 - 2023-09-18 - Metrics data model refactor
+
+Vector's data model for metrics incorporates a large number of possible metric types that it can
+natively handle, along with a variety of information/metadata about those metrics. Over time, easily
+(and correctly) handling all of those metric types within each metrics-specific source and sink
+component has been difficult, consuming both a large amount of developer time as well as leading to
+subtle and hard-to-diagnose issues when shuttling data from one metrics system/encoding to another.
+This RFC proposes a body of refactoring work meant to addresses current issues with the data model,
+as well as help future-proof it.
+
+## Context
+
+- [Support AggregatedSummary for `gcp_stackdriver` sink](https://github.com/vectordotdev/vector/issues/9530)
+- [Support aggregated histograms in `statsd` sink](https://github.com/vectordotdev/vector/issues/11661)
+- [Prometheus remote write source has incomplete support/behavior.](https://github.com/vectordotdev/vector/issues/10801)
+- [Cumulative histogram in aggregation transform](https://github.com/vectordotdev/vector/issues/12745)
+- [Support Prometheus native histograms](https://github.com/vectordotdev/vector/issues/16333)
+- [Support converting Histogram to Summary](https://github.com/vectordotdev/vector/issues/13754)
+
+## Cross-cutting concerns
+
+- Stability / schema guarantees around `native_json` codec by basing it on the Protocol Buffers
+  schema (no overarching GitHub issue for this yet) and how changes proposed by this RFC would alter
+  the output of that codec
+
+## Scope
+
+### In scope
+
+- Adding/removing metric types to the data model.
+- Updating metrics-capable components to handle changes to the metrics data model.
+
+### Out of scope
```

Comment:
It seems like we're not revisiting `MetricKind` and our concept of incremental vs absolute here, which is fine, but it might be worth noting that and putting at least a small discussion of the topic in a future work section. It does still seem like there will be some odd possibilities (e.g. "absolute" raw distribution) after this work is complete, which would be nice to eventually eliminate.

Reply:
Is the counter metric the only one for which the absolute vs incremental distinction makes sense?

Reply:
Maybe? And if the intent of an absolute counter is "don't aggregate me", then is it semantically equivalent to a gauge? I'm not sure.

Reply:
OpenTelemetry handles this by bundling the aggregation temporality (our `MetricKind`) with the metric type (our `MetricValue`) itself. Counters ("sums" in OTel) _do_ have an aggregation temporality, along with (exponential) histograms, but gauges do not.

Reply:
Addressed in 7b92f3684.

I added an out-of-scope item here _and_ a corresponding item under "Future Improvements".